Fix picker position

tungi52 opened the pull request:
Possible fix to #267 

@Garderoben @henon - a new class `BoundingClientRect` was necessary for JS interop, but I'm not sure what is the correct place for these objects. For now, I have created a new folder.

mikes-gh commented:
This looks good.  You have some failing tests. 😄
ClientBoundingRect looks like a useful utility function and I can see it being used in other places.
It might be worth creating a IClientBoundingRectProvider?
var clientRect = await ClientBoundingRect.FromRef(PickerContainerRef);

mikes-gh commented:
And you're right it should be more obvious where to put the utility classes.  I am going to look at the namespaces/folder structure when I tidy up the MudBlazor.Services namespace

mikes-gh commented:
👍 bunit chnaged recently

bUnit's mock IJSRuntime has been moved to an "always on" state by default, in strict mode, and is now available through TestContext's JSInterop property. This makes it possible for first party Blazor components like the <Virtualize> component, which depend on JSInterop, to "just work" in tests.

mikes-gh commented:
If it doesnt work you may need to use 

Compatible with previous releases: To get the same effect as calling Services.AddMockJSRuntime() in beta-11, which used to add the mock IJSRuntime in "loose" mode, you now just need to change the mode of the already on JSInterop, i.e. ctx.JSInterop.Mode = JSRuntimeMode.Loose.

tungi52 commented:
Thank you, I check it.

mikes-gh commented:
Me again.  Random thoughts this maybe for another commit but the user may need some control with a PopupPosition.
I would consider this PR to be the Auto setting.  User can override with Top Bottom Left Right?

mikes-gh commented:
You shouldnt call ctx.Services.AddMockJSRuntime(); its done for you also the using statement should be `using Bunit.TestDoubles` If you still have trouble try ctx.JSInterop.Mode = JSRuntimeMode.Loose

tungi52 commented:
Thank you! I added ctx.Services.AddMockJSRuntime(); because the tests on my computer failed. I removed it, and interestingly now it is working. Maybe I had other problems that confused me and mixed up things.

tungi52 commented:
Yout thought about the positioning is interesting. I can imagine only left-right-center, because of the current viewport problem, I think the vertical positioning should be auto.

mikes-gh commented:
I would say you need an override on the vertical.  You may be tight on space and prefer the user to scroll a bit.  BUt I would say its for a seperate PR

mikes-gh commented:
I see the problem your develop is out of date and bunit was updated recently.
Please update develop and rebase

mikes-gh commented:
Also you there are API tests failing due to the new dependency of MudBlazorResizeListener service. @henon How expensive perfomance wise is this service?

mikes-gh commented:
Well done on working out the tests.  There are some PRs in the pipeline that update the API test setup
Can you rebase and squash please.

tungi52 commented:
@mikes-gh I'm not a git expert, so I'm not sure I made it right.

mikes-gh commented:
@tungi52 what IDE are you using?

tungi52 commented:
VS 2019. For git operations I use TortoiseGit and Github Desktop.
